### PR TITLE
MN pupa scraper

### PR DESCRIPTION
**Pull request by divergentdave**
This is a rewrite of the MN scraper in the pupa framework. It's very much a work in progress; right now only the legislators are done. This depends on the pupa-related infrastructure added in #1399.

### CODE REVIEW

**Comment by divergentdave** [comment]
I'm working on this some more this weekend. Two notes for later review:

* I'm using the master branch of the spatula library for now. (to pull in jamesturk/spatula#1) Before this is finalized, a new version will need to be pushed to PyPI, I'll unwind the changes made here to requirements-new.txt and docker-compose.yml, and a new Docker image will need to be pushed to dockerhub.

* In the bill scraper, one of the `bill.add_version(` calls set `on_duplicate='use_new'`. Per the docs, I've tentatively replaced that with `on_duplicate='ignore'`, but the docs mention this probably needs discussion here.

**Comment by divergentdave** [comment]
Furthermore, it appears that the MN scraper was the only one using `bill['scraped_subjects']`. A comment says:

    Currently we are not mapping to Open States standardized subjects, so use 'scraped_subjects'

I've commented this out for now. I assume at some point this mapping to standardized subjects will need to be done.

**Comment by divergentdave** [comment]
This is working now! I'll be reviewing it a little closer later in the week for data quality and code issues.

@jamesturk could you cut a new release of spatula and push an updated image of openstates-pupa?

Also I'd appreciate expert input on the `on_duplicate` and `scraped_subjects` questions above from anyone.

Thanks!

**Comment by jamesturk** [comment]
awesome, and I'll start taking a look at this

File: openstates/mn/__init__.py [hunk 1]
```diff
@@ -1,10 +1,11 @@
-import lxml.html
-import gc
+# encoding=utf-8
```
Comment (jmcarp) [inline openstates/mn/__init__.py]: This is the default source code encoding in python3, so I don't think we need it.

File: docker-compose.yml [hunk 2]
```diff
@@ -21,7 +21,9 @@ services:
     depends_on:
     - database
   scrape:
-    image: openstates/openstates-pupa
+    build:
```
Comment (jmcarp) [inline docker-compose.yml]: I'm guessing this was for testing local changes--did you want this change to be part of the patch?

File: requirements-new.txt [hunk 3]
```diff
@@ -3,4 +3,4 @@ xlrd
 lxml>=2.2
 pytz>=2010l
 feedparser>=4.1
-spatula
+git+https://github.com/jamesturk/spatula.git
```
Comment (jmcarp) [inline requirements-new.txt]: Looks like spatula 0.0.2 just dropped! 🎉

File: openstates/mn/bills.py [hunk 4]
```diff
@@ -1,65 +1,85 @@
...
+    '2013-2014': '0882013',
+    '2013s1': '1882013',
+    '2015-2016': '0892015',
+    '2015s1': '1892015',
+    '2017-2018': '0902017',
+}
+
+
+class MNBillScraper(Scraper, LXMLMixin):
     jurisdiction = 'mn'
```
Comment (jmcarp) [inline openstates/mn/bills.py]: Can we drop this?

File: openstates/mn/bills.py [hunk 5]
```diff
@@ -157,47 +187,59 @@ def get_bill_info(self, chamber, session, bill_detail_url, version_list_url):
...
+                    title=bill_title, classification=bill_type)
 
         # Add source
         bill.add_source(bill_detail_url)
 
         # Add subjects.  Currently we are not mapping to Open States
         # standardized subjects, so use 'scraped_subjects'
-        bill['scraped_subjects'] = self._subject_mapping[bill_id]
+        # TODO: migrate to Pupa
+        # bill['scraped_subjects'] = self._subject_mapping[bill_id]
```
Comment (jmcarp) [inline openstates/mn/bills.py]: @jamesturk: I see that billy defines a standard list of subjects, but I don't see anything comparable in pupa / ocd (just took a cursory look though). Are we keeping a distinction between `subjects` and `scraped_subjects` going forward?

File: openstates/mn/__init__.py (hunk 1 above)
Comment (divergentdave) [inline openstates/mn/__init__.py]: :+1:

File: requirements-new.txt (hunk 3 above)
Comment (divergentdave) [inline requirements-new.txt]: Excellent, backing this out.

File: openstates/mn/bills.py (hunk 5 above)
Comment (jamesturk) [inline openstates/mn/bills.py]: correct- we don't have this distinction anymore, you can use bill.add_subject() to add these directly

File: docker-compose.yml (hunk 2 above)
Comment (divergentdave) [inline docker-compose.yml]: This and the requirements-new.txt changes were for testing purposes. The image isn't up to date yet, but now that spatula 0.0.2 is out, the next commit to this repository should fix that. I'll back this out in anticipation of the next automated build.

File: openstates/mn/bills.py (hunk 4 above)
Comment (divergentdave) [inline openstates/mn/bills.py]: Good catch, done

File: openstates/mn/bills.py (hunk 5 above)
Comment (divergentdave) [inline openstates/mn/bills.py]: :+1: done

File: openstates/mn/people.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+    def handle_list_item(self, item):
+        photo_url = item.xpath('./td[1]/a/img/@src')[0]
+        info_nodes = item.xpath('./td[2]/p/a')
+        name_text = info_nodes[0].xpath('./b/text()')[0]
+        url = info_nodes[0].get('href')
+
+        name_match = re.match(r'^(.+)\(([0-9]{2}[AB]), ([A-Z]+)\)$', name_text)
+        name = name_match.group(1).strip()
+        district = name_match.group(2).lstrip('0').lower()
```
Comment (jamesturk) [inline openstates/mn/people.py]: can we switch this to upper() to match districts

File: openstates/mn/__init__.py [hunk 7]
```diff
@@ -161,23 +120,29 @@
...
 
-def session_list():
-    from billy.scrape.utils import url_xpath
-    return url_xpath('https://www.revisor.mn.gov/revisor/pages/search_status/'
-                     'status_search.php?body=House',
-                     '//select[@name="session"]/option/text()')
+        for n in range(1, 68):
+            upper.add_post(label=str(n), role='Senator',
+                           division_id='ocd-division/country:us/state:mn/sldu:{}'.format(n))
+            lower.add_post(label=str(n) + 'a', role='Representative',
```
Comment (jamesturk) [inline openstates/mn/__init__.py]: could you change the label to be upper case to match existing Open States district naming

**Comment by jamesturk** [comment]
from what I can tell from comparing output this is almost good to go 👍

File: openstates/mn/people.py (hunk 6 above)
Comment (divergentdave) [inline openstates/mn/people.py]: :+1: done

File: openstates/mn/__init__.py (hunk 7 above)
Comment (divergentdave) [inline openstates/mn/__init__.py]: Okay, did that. For clarity, we're leaving the ocd division alone, right? i.e. `ocd-division/country:us/state:mn/sldl:1a`

Comment (jamesturk) [inline openstates/mn/__init__.py]: yep, looks good!

**Comment by divergentdave** [comment]
From my testing it looks like the events scraper needs some more work. Currently it's missing a lot of meetings where the word "Room" is not bolded and not followed by a colon. Additionally, when a body is in session, that's getting picked up as "Live Video HTV 1".

From discussion above, could someone take a look at the `on_duplicate` argument to `bill.add_version()`? Does this need to be changed at all?

**Comment by jamesturk** [comment]
I think the on_duplicate stuff should be fine 👍

**Comment by divergentdave** [comment]
I fixed the event scraper in a couple more follow-on commits. This should be good to go!

**Comment by divergentdave** [comment]
@jamesturk @mileswwatkins ^ ping. This scraper is ready to go.

**Comment by jamesturk** [comment]
cool, this is next in my queue to do one final review on then merge.  I'm a bit slammed right now & want to be careful w/ merging pupa ports which is why we're a bit backed up right now but we'll be tackling these very soon 👍

**Comment by jamesturk** [comment]
OK, after a bit of a pause this looks like it had fixed all issues I had noted. switching it over!

**Pull request merged**